plugins/out_stackdriver/stackdriver.c

```diff
@@ -2498,14 +2502,26 @@ static void cb_stackdriver_flush(struct flb_event_chunk *event_chunk,
         flb_upstream_conn_release(u_conn);
         FLB_OUTPUT_RETURN(FLB_RETRY);
     }
-    payload_size = flb_sds_len(payload_buf);
+
+    out_payload_buf = (flb_sds_t) payload_buf;
+    out_payload_size = flb_sds_len(payload_buf);
+    if (ctx->compress != NULL && strcmp(ctx->compress, "gzip") == 0) {
+        ret = flb_gzip_compress((void *) payload_buf, flb_sds_len(payload_buf),
+                                &out_payload_buf, &out_payload_size);
+        if (ret == -1) {
+            flb_error("[out_stackdriver] cannot gzip payload, disabling compression");
+        } else {
+            compressed = FLB_TRUE;
+            flb_sds_destroy(payload_buf);
+        }
+    }
 
     /* Get or renew Token */
     token = get_google_token(ctx);
     if (!token) {
         flb_plg_error(ctx->ins, "cannot retrieve oauth2 token");
         flb_upstream_conn_release(u_conn);
-        flb_sds_destroy(payload_buf);
+        flb_sds_destroy(out_payload_buf);
```

Comment:
Please revert the changes related to `payload_buf`, add a third variable to hold the final payload (and one for its sice) that's conditionally set to either a copy the raw result of `stackdriver_format` or the output buffer `flb_gzip_compress` produces to make the usage transparent while retaining a proper cleanup procedure.